sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RetryPolicy.java

```diff
@@ -112,56 +119,97 @@ public RetryPolicy(RetryStrategy retryStrategy) {
+    }
+
+    /*
      * Determines the delay duration that should be waited before retrying.
-     *
-     * @param response HTTP response
-     * @return If the HTTP response has a retry-after-ms header that will be returned, otherwise the duration used
-     * during the construction of the policy.
      */
-    private Duration determineDelayDuration(HttpResponse response, int tryCount) {
-        int code = response.getStatusCode();
+    static Duration determineDelayDuration(HttpResponse response, int tryCount, RetryStrategy retryStrategy,
+        String retryAfterHeader, ChronoUnit retryAfterTimeUnit) {
+        // If the retry after header hasn't been configured, attempt to look up the well-known headers.
+        if (isNullOrEmpty(retryAfterHeader)) {
+            return getWellKnownRetryDelay(response.getHeaders(), tryCount, retryStrategy);
+        }
+
+        String retryHeaderValue = response.getHeaderValue(retryAfterHeader);
 
-        // Response will not have a retry-after-ms header.
-        if (code != 429        // too many requests
-            && code != 503) {  // service unavailable
+        // Retry header is missing or empty, return the default delay duration.
+        if (isNullOrEmpty(retryHeaderValue)) {
             return retryStrategy.calculateRetryDelay(tryCount);
         }
 
-        String retryHeaderValue = null;
+        // Use the response delay duration, the server returned it for a reason.
+        return Duration.of(Integer.parseInt(retryHeaderValue), retryAfterTimeUnit);
+    }
 
-        if (!isNullOrEmpty(this.retryAfterHeader)) {
-            retryHeaderValue = response.getHeaderValue(this.retryAfterHeader);
+    /*
+     * Determines the delay duration that should be waited before retrying using the well-known retry headers.
+     */
+    static Duration getWellKnownRetryDelay(HttpHeaders responseHeaders, int tryCount, RetryStrategy retryStrategy) {
+        String retryHeaderValue = responseHeaders.getValue(X_MS_RETRY_AFTER_MS_HEADER);
+
+        // Found 'x-ms-retry-after-ms' header, use a Duration of milliseconds based on the value.
+        if (!isNullOrEmpty(retryHeaderValue)) {
+            int delayMs = tryParseHeaderToInt(retryHeaderValue);
+            if (delayMs != Integer.MIN_VALUE) {
+                return Duration.ofMillis(delayMs);
+            }
         }
 
-        // Retry header is missing or empty, return the default delay duration.
-        if (isNullOrEmpty(retryHeaderValue)) {
-            return this.retryStrategy.calculateRetryDelay(tryCount);
+        retryHeaderValue = responseHeaders.getValue(RETRY_AFTER_MS_HEADER);
+
+        // Found 'retry-after-ms' header, use a Duration of milliseconds based on the value.
+        if (!isNullOrEmpty(retryHeaderValue)) {
+            int delayMs = tryParseHeaderToInt(retryHeaderValue);
+            if (delayMs != Integer.MIN_VALUE) {
+                return Duration.ofMillis(delayMs);
+            }
```

Comment:
feel this could be put seperate in the helper method instead of the method `tryParseHeaderToInt`